aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/MilestoneStepTest.java

```diff
@@ -0,0 +1,166 @@
+import org.junit.Test;
+import org.junit.runners.model.Statement;
+import org.jvnet.hudson.test.BuildWatcher;
+import org.jvnet.hudson.test.RestartableJenkinsRule;
+
+import static org.junit.Assert.*;
+
+import hudson.model.Result;
+
+public class MilestoneStepTest {
+
+    @ClassRule
+    public static BuildWatcher buildWatcher = new BuildWatcher();
+
+    @Rule
+    public RestartableJenkinsRule story = new RestartableJenkinsRule();
+
+    @Before
+    public void clear() {
+        MilestoneStepExecution.clear();
+    }
+
+    @Test
+    public void buildsMustPassThroughInOrder() {
+        story.addStep(new Statement() {
+            @Override public void evaluate() throws Throwable {
+                WorkflowJob p = story.j.jenkins.createProject(WorkflowJob.class, "p");
+                p.setDefinition(new CpsFlowDefinition(
+                        "semaphore 'inorder'\n" +
+                        "echo 'Before milestone'\n" +
+                        "milestone()\n" +
+                        "echo 'Passed first milestone'\n" +
+                        "milestone()\n" +
+                        "echo 'Passed second milestone'\n" +
+                        "milestone()\n" +
+                        "echo 'Passed third milestone'\n"));
+                WorkflowRun b1 = p.scheduleBuild2(0).waitForStart();
+                SemaphoreStep.waitForStart("inorder/1", b1);
+                WorkflowRun b2 = p.scheduleBuild2(0).waitForStart();
+                SemaphoreStep.waitForStart("inorder/2", b2);
+
+                // Let #2 continue so it finish before #1
+                SemaphoreStep.success("inorder/2", null);
+                story.j.assertBuildStatusSuccess(story.j.waitForCompletion(b2));
+
+                // Let #1 continue, so it must be early cancelled since #2 already went into milestone 1
+                SemaphoreStep.success("inorder/1", null);
+                story.j.assertBuildStatus(Result.NOT_BUILT, story.j.waitForCompletion(b1));
+                story.j.assertLogNotContains("Passed first milestone", b1);
+            }
+        });
+    }
+
+    @Test
+    public void olderBuildsMustBeCancelledOnMilestoneExit() {
+        story.addStep(new Statement() {
+            @Override public void evaluate() throws Throwable {
+                WorkflowJob p = story.j.jenkins.createProject(WorkflowJob.class, "p");
+                p.setDefinition(new CpsFlowDefinition(
+                        "milestone()\n" +
+                        "echo 'First milestone'\n" +
+                        "semaphore 'wait'\n" +
+                        "milestone()\n" +
+                        "echo 'Second milestone'\n"));
+                WorkflowRun b1 = p.scheduleBuild2(0).waitForStart();
+                SemaphoreStep.waitForStart("wait/1", b1);
+                WorkflowRun b2 = p.scheduleBuild2(0).waitForStart();
+                SemaphoreStep.waitForStart("wait/2", b2);
+                // Now both #1 and #2 are into milestone 1
```

Comment:
I find it easier to think, and write, in terms of “passing” rather than “entering”. They are between M1 and M2.